core/src/main/java/org/apache/calcite/rel/type/RelDataTypeSystemImpl.java

```diff
@@ -255,8 +268,72 @@ && getDefaultPrecision(typeName) != -1) {
     return argumentType;
   }
 
-  @Override public RelDataType deriveCovarType(RelDataTypeFactory typeFactory,
+  @Override public RelDataType deriveCovarType(RelDataTypeFactory typeFactory, SqlKind sqlKind,
       RelDataType arg0Type, RelDataType arg1Type) {
+    switch (sqlKind) {
```

Comment:
I'm not sure if this is a reasonable implementation, but from the test cases, it seems feasible.